Fiddler receiver basic implementation

samikshan-fiddler opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This is the first basic implementation for the Fiddler receiver. Through this implementation, the receiver can be configured to against any fiddler deployment through the fiddler endpoint and authentication token to consume metrics data for all models onboarded with Fiddler and converts them to OTLP gauge metric type.

We have also added another option to configure the set of metrics we want to consume data for through the receiver. The default interval for the receiver to run is `3600s` i.e. 1 hour.

samikshan-fiddler committed:
Added basic implementation for the fiddler receiver

Fetches metrics data from Fiddler all at regular configurable
intervals for all onboarded models and converts to OTLP gauge
metric type.

samikshan-fiddler committed:
Log enabled metrics

samikshan-fiddler committed:
Retain timestamps from Fiddler timeseries data

samikshan-fiddler committed:
Fix metrics collection on each configurable interval

Fix metrics collection on each configurable interval by directly
collecting metrics at each ticker interval without additional time tracking.

samikshan-fiddler committed:
Fix issue with resolving timestamp format

The timestamp data was being incorrectly parsed. This fixes
the logic that parses the timestamp from the metrics data as
well as updates the tests and the sample response data
accordingly

samikshan-fiddler committed:
Update transformation of Fiddler metric to OTLP

- Reuse metrics data available per model to get the metric
types and use in OTLP metric name
- Update the tests with relevant response format for metrics
data retrieved through the `POST /v3/queries` API call
- Ignore metrics that require categories.

samikshan-fiddler committed:
Updated defaults and test config for fiddler receiver

samikshan-fiddler committed:
Improved error handling and logging

samikshan-fiddler committed:
Fiddler receiver config update

Update parameter to configure enabled metric types, from
`enabledMetrics` to `enabled_metric_types`

samikshan-fiddler committed:
Remove metric type validation in Fiddler receiver config

There is no need to verify metric type validation in fiddler receiver config.
We check the enabled metric types configured against the metric types
available per model anyway. Having the metric type validation on config
validation stage makes the process more error prone and hardcoded if
anything.

samikshan-fiddler committed:
Fix logs against update configuration name

samikshan-fiddler committed:
Added debug logs for queries in `POST v3/queries` API call

samikshan-fiddler committed:
Do a collection immediately after the receiver starts

Currently, for the first collection of metrics, we need to wait for
the configurable interval amount. This changes the behaviour
to ensure that we do a collection immediately after the receiver
starts and then waits for the configured interval before the next
collection of metrics data from Fiddler

samikshan-fiddler committed:
Add model version as data point attribute

samikshan-fiddler committed:
Ensure that timestamps are in UTC when converting to OTLP

samikshan-fiddler committed:
Skip any violations data integrity metrics

Add a temporary workaround to skip specific data integrity metrics
with known API handling issues. This prevents potential errors when
collecting metrics for models using the Fiddler receiver.

samikshan-fiddler committed:
Test E2E collection behaviour for fiddler receiver

robin-fiddler commented on receiver/fiddlerreceiver/receiver.go:
```diff
@@ -5,23 +5,292 @@ package fiddlerreceiver // import "github.com/open-telemetry/opentelemetry-colle
+	mb := metadata.NewMetricBuilder(fr.logger)
+
+	// Process each model
+	for _, model := range models {
```
This would be sequential and synchronous. Will this be able to collect all metrics within the stipulated time frame?

samikshan-fiddler replied:
Given a minimum interval time of 5 minutes, my estimate is that this receiver will start facing issues once we hit situations where we are dealing with 100+ models. (if we consider a maximum of 3s latency per query). We can fetch the metrics data in parallel and then merge them through channels to improve this significantly. I will work on this on a follow up PR as well.

robin-fiddler commented on receiver/fiddlerreceiver/internal/client/client.go:
```diff
@@ -0,0 +1,247 @@
+	httpClient := &http.Client{
+		Timeout: cfg.timeout,
+	}
+
```
using connection pooling may improve performance. similar to the fiddler monorepo, we could consider a retry policy for retryable failures.
https://pkg.go.dev/net/http#hdr-Clients_and_Transports

samikshan-fiddler replied:
I will add both explicit connection pooling and retries on retryable failures (for example, 5xx responses) as follow up enhancements. Go http connections already has connection pooling by default (idle connection timeout of 90 seconds with a pool size of 100 idle connections) that can handle initial usecases.

samikshan-fiddler committed:
Merge branch 'downstream-add-fiddler-receiver' into fiddler-receiver-basic-implementation

robin-fiddler reviewed (approved):
LGTM. Agreed that enhancements suggested in comments can be taken up as a follow up PR.

samikshan-fiddler committed:
Update Makefile include path for fiddlereceiver

samikshan-fiddler committed:
Update Makefile

samikshan-fiddler committed:
Add copyright and license headers to Fiddler receiver files

samikshan-fiddler committed:
Merge branch 'fiddler-receiver-basic-implementation' of github.com:fiddler-labs/opentelemetry-collector-contrib into fiddler-receiver-basic-implementation

Ramjee-Fiddler commented on receiver/fiddlerreceiver/internal/client/client.go:
```diff
@@ -0,0 +1,247 @@
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.token))
+	req.Header.Set("Content-Type", "application/json")
```
Can you add another header similar to our data dog integration.

https://github.com/DataDog/integrations-extras/blob/3e8359bd408077e46503b57e9a115778f155f1d6/fiddler/datadog_checks/fiddler/check.py#L43

It will be something like:

'X-Integration': 'fiddler-otel',

samikshan-fiddler replied:
@Ramjee-Fiddler Sure. Is this something that will enable us to differentiate calls to the `/v3/queries` endpoint from this receiver vs UI, when we observe for latency against this endpoint?

samikshan-fiddler replied:
Have added this to the header.

samikshan-fiddler committed:
Merge branch 'development' into fiddler-receiver-basic-implementation

samikshan-fiddler committed:
Merge branch 'development' into fiddler-receiver-basic-implementation

samikshan-fiddler committed:
Update fiddler receiver implementation

- updated usage of `receivertest.NewNopSettings()`
- Updated `go.mod` and `go.sum`

samikshan-fiddler committed:
Add X-Integration header for HTTP requests to Fiddler endpoints

Set the X-Integration header as 'fiddler-otel-receiver' in the
HTTPClient call method to ensure better request tracking and
integration identification for calls to Fiddler API endpoints.
